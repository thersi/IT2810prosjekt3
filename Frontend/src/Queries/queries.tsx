import { gql } from "@apollo/client";

export const QUERY_MOVIE_BY_ID = gql`
  query ($movieByIdId: ID!) {
    movieById(id: $movieByIdId) {
      _id
      title
      year
      thumbsUp
      thumbsDown
      genre
      actors
      poster
    }
  }
`;

export const QUERY_ALL_MOVIES = gql`
  query (
    $filterGenre: String!
    $limit: Int!
    $page: Int!
    $order: Int!
    $sortOn: String!
    $word: String!
  ) {
    searchAndFilter(
      filterGenre: $filterGenre
      limit: $limit
      page: $page
      order: $order
      sortOn: $sortOn
      word: $word
    ) {
      _id
      title
      year
      thumbsUp
      thumbsDown
      poster
    }
  }
`;

export const THUMBS_UP_MUTATION = gql`
  mutation ($thumbsUpByIdId: ID!) {
    thumbsUpById(id: $thumbsUpByIdId) {
      thumbsUp
    }
  }
`;
export const THUMBS_DOWN_MUTATION = gql`
  mutation ($thumbsDownByIdId: ID!) {
    thumbsDownById(id: $thumbsDownByIdId) {
      thumbsDown
    }
  }
`;
