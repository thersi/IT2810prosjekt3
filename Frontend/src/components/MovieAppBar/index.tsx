import AppBar from "@mui/material/AppBar";
import Box from "@mui/material/Box";
import Toolbar from "@mui/material/Toolbar";
import Typography from "@mui/material/Typography";
import SearchBar from "../SearchBar";
import MovieFilter from "../MovieFilter";
import { MovieAppBarProps } from "../../Interfaces/Interfaces";

export default function MovieAppBar(props: MovieAppBarProps) {
  //The below methods are used to get parent from child to parent
  let searchValue = (value: string) => {
    props.handleSearch(value);
  };
  let filterValue = (value: string) => {
    props.handleFilter(value);
  };
  let sortValue = (value: boolean) => {
    props.handleSort(value);
  };
  return (
    <Box sx={{ flexGrow: 1 }}>
      <AppBar position="static">
        <Toolbar>
          <Typography
            variant="h6"
            noWrap
            component="div"
            sx={{ flexGrow: 1, display: { xs: "none", sm: "block" } }}
          >
            MÃ¸vies
          </Typography>
          <MovieFilter
            aria-label={
              "Movie filter to filter movies by ascending or descending order. Also filters by title og year. "
            }
            aria-required="true"
            handleFilter={filterValue} //Sends filter value from child to parent
            handleSort={sortValue} //Sends sort value from child to parent
          />
          <SearchBar
            aria-label={"Search bar to search through movie titles"}
            aria-required="true"
            handleSearch={searchValue} //Sends search value from child to parent
          />
        </Toolbar>
      </AppBar>
    </Box>
  );
}
