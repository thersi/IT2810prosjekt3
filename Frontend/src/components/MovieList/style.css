.movieContainer {
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  gap: 10px;
  align-content: center;
  justify-content: center;
  padding-left: 5%;
  padding-right: 5%;
}

