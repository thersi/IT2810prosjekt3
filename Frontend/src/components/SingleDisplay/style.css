
.movie {
  display: flex;
  padding: 3px;
  border: 2px solid;
  border-color: rgba(46, 140, 240, 0.5);
  background-color: rgba(46, 140, 240, 0.2);
  width: 80%;
  font-family: Arial, Helvetica, sans-serif;
}
@media screen and (max-width: 600px) {
  .movie {
    width: 95%;
  }
}

.movieInfo {
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-left: 3px;
}

img {
  text-align: center;
  flex-basis: 40%;
  display: block;
  width: 120px;
  height: 160px;
}

.text {
  padding-left: 2%;
  font-size: 20px;
  flex-direction: column;
  display: block;
  width: 100%;
  height: 100%;
  padding-top: 3%;
}

.thumbButtons {
  display: flex;
  flex-direction: flex-end;
}
.thumbUp {
  padding: 20pt;
  color: rgb(106, 141, 106);
}
.thumbDown {
  padding: 20pt;
  color: rgb(194, 122, 122);
}
