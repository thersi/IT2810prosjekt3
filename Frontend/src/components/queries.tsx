import React, { useState } from "react";
import { useQuery, gql } from "@apollo/client";

//DETTE ER KUN SELVE QUERIENE. INGEN HÃ…NDTERING ELLER LOGIKK
//FLYTT DEM DIT DE TRENGS OG BRUK USEQUERY

interface Movie {
  _id: string;
  title: string;
  thumbsUp: number;
  year: number;
  genere: string[];
  actors: string[];
  thumbsDown: number;
  poster: string;
}

/* gets all movies (gets only specified attributes).  
limit: number of elements to load
page: determine offset (ex: first three, next three, nextnext three etc.
order: 1 (ASC) or -1 (DESC). 
sortOn: "title" or "year"
*/
const QUERY_ALL_MOVIES = gql`
  query ($limit: Int!, $page: Int!, $order: Int!, $sortOn: String!) {
    movies(limit: $limit, page: $page, order: $order, sortOn: $sortOn) {
      _id
      title
      year
      thumbsUp
      thumbsDown
    }
  }
`;
//CAN FILL INN MORE ATTRIBUTES FROM INTERFACE Movie TO GET MORE INFO

const QUERY_ALL_FILTERED_GENRE = gql`
  query (
    $filterGenre: String!
    $limit: Int!
    $page: Int!
    $order: Int!
    $sortOn: String!
  ) {
    filterOnGenre(
      filterGenre: $filterGenre
      limit: $limit
      page: $page
      order: $order
      sortOn: $sortOn
    ) {
      _id
      title
      year
      thumbsUp
      thumbsDown
    }
  }
`;

const QUERY_ALL_CONTAINS_SEARCH = gql`
    query($limit: Int!, $page: Int!, $word: String!, $order: Int!, $sortOn: String!){
        containsString(limit: $limit, page: $page, word: $word, order: $order, sortOn: $sortOn) {
            _id
            title
            year
            thumbsUp
            thumbsDown
        }
    }
}
`;

const QUERY_MOVIE_BYID = gql`
    query($movieByIdId: ID!){
        movieById(id: $movieByIdId) {
            title
            year
            thumbsUp
            thumbsDown
        }
    }
}
`;

const QUERY_MOVIE_BYTITLE = gql`
    query($title: String!){
        movieByTitle(title: $title) {
            title
            year
            thumbsUp
            thumbsDown
        }
    }
}
`;
