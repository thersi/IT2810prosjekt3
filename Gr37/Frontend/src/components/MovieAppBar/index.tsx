import AppBar from "@mui/material/AppBar";
import Box from "@mui/material/Box";
import Toolbar from "@mui/material/Toolbar";
import IconButton from "@mui/material/IconButton";
import Typography from "@mui/material/Typography";
import MenuIcon from "@mui/icons-material/Menu";
import SearchBar from "../SearchBar";
import MovieFilter from "../MovieFilter";
import { createTheme } from "@mui/system";

export default function MovieAppBar() {
  const darkTheme = createTheme({
    palette: {
      mode: "dark",
    },
  });
  return (
    <Box sx={{ flexGrow: 1 }}>
      <AppBar position="static">
        <Toolbar>
          <Typography
            variant="h6"
            noWrap
            component="div"
            sx={{ flexGrow: 1, display: { xs: "none", sm: "block" } }}
          >
            Movies
          </Typography>
          <MovieFilter />
          <SearchBar />
        </Toolbar>
      </AppBar>
    </Box>
  );
}
