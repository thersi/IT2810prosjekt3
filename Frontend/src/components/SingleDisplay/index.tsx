import { useState } from "react";
import MovieDialog from "../MovieDialog";
import "./style.css";
import ThumbUpIcon from "@material-ui/icons/ThumbUp";
import ThumbDownIcon from "@material-ui/icons/ThumbDown";
import {
  MovieByIdInput,
  MovieByIdResult,
  SingleDisplayProps,
} from "../../Interfaces/Interfaces";
import { useLazyQuery } from "@apollo/client";
import { QUERY_MOVIE_BY_ID } from "../../Queries/queries";

export default function SingleDisplay(props: SingleDisplayProps) {
  const { movie, refetch } = props;
  const [open, setOpen] = useState(false);
  const [thumbsUp, setThumbsUp] = useState(movie.thumbsUp);
  const [thumbsDown, setThumbsDown] = useState(movie.thumbsDown);

  const vote = sessionStorage.getItem(movie._id);
  const dflt_vote = vote !== null ? JSON.parse(vote) : false;
  const [voted, setVoted] = useState(dflt_vote);

  const [fetchMovie, { data: movieByIdData, loading: movieByIdLoading }] =
    useLazyQuery<MovieByIdResult, MovieByIdInput>(QUERY_MOVIE_BY_ID);

  if (movieByIdLoading) {
    return <p>Movie is loading</p>;
  }

  if (open && typeof movieByIdData !== "undefined") {
    return (
      <MovieDialog
        movie={movieByIdData.movieById}
        setOpen={setOpen}
        setThumbsUp={setThumbsUp}
        thumbsUp={thumbsUp}
        setThumbsDown={setThumbsDown}
        thumbsDown={thumbsDown}
        voted={voted}
        setVoted={setVoted}
        refetch={refetch}
      />
    );
  }

  return (
    <div
      className="movie"
      onClick={() => {
        setOpen(true);
        fetchMovie({ variables: { movieByIdId: movie._id } });
      }}
    >
      <div>
        <img className="image" src={movie.poster} alt="movie"></img>
      </div>
      <div className="movieInfo">
        <div className="text">
          {<label>{movie.title + " (" + movie.year + ") "}</label>}
        </div>
        <div className="thumbButtons">
          {" "}
          <div className="thumbUp">
            <ThumbUpIcon />
            {thumbsUp}
          </div>
          <div className="thumbDown">
            <ThumbDownIcon />
            {thumbsDown}
          </div>
        </div>
      </div>
    </div>
  );
}
