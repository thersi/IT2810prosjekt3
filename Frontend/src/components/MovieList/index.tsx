import { Movie, MovieListProps } from "../../Interfaces";
import SingleDisplay from "../SingleDisplay";
import QueryMoviesResult from "../MovieSearch";
import "./style.css";

const MovieList = (props: MovieListProps) => {
  /* const { loadingTitle, errorTitle, dataTitle } = useQuery(
    QUERY_MOVIE_BYTITLE,
    {
      variables: { title: search },
    }
  ); */
  const { data } = props;

  return (
    <div className="movieContainer">
      {props.data.searchAndFilter.map((movie: Movie) => (
        <SingleDisplay {...movie} />
      ))}
    </div>
  );
};

export default MovieList;
