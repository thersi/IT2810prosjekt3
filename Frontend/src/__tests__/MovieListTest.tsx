import EnzymeToJson from "enzyme-to-json";
import { configure, render, shallow } from "enzyme";
import MovieAppBar from "../components/MovieAppBar";
import Adapter from "enzyme-adapter-react-16";
import MovieList from "../components/MovieList";
import { useQuery } from "@apollo/client";
import {
  Movie,
  QueryMoviesInput,
  QueryMoviesResult,
} from "../Interfaces/Interfaces";
import { QUERY_ALL_MOVIES } from "../Queries/queries";
import SingleDisplay from "../components/SingleDisplay";

const createMovie = (id: string) => {
  const movie: Movie = {
    _id: id,
    title: "Avengers: Endgame",
    year: 2019,
    thumbsUp: 196,
    thumbsDown: 37,
    genre: ["action", "comedy"],
    actors: ["Robert Downey Jr.", "Chris Evans"],
    poster: "url",
  };
  return movie;
};
const data: Movie[] = [];
for (let i = 0; i < 10; i++) {
  data.push(createMovie(i.toString()));
}

const component = <MovieList data={data} />;

configure({ adapter: new Adapter() });
describe("Movie App Bar", () => {
  it("renders correctly", () => {
    render(component);
  });
  it("Should match snapshot", () => {
    const snapshotCheck = shallow(component);
    expect(EnzymeToJson(snapshotCheck)).toMatchSnapshot();
  });

  it("Should display correct number of movies", () => {
    const shallowComp = shallow(component);
    expect(shallowComp.find(SingleDisplay)).toHaveLength(10);
  });
});
