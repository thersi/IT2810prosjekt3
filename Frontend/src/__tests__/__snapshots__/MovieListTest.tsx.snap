// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie App Bar Should match snapshot 1`] = `
<div
  className="movieContainer"
>
  <SingleDisplay
    key="0"
    movie={
      Object {
        "_id": "0",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
  <SingleDisplay
    key="1"
    movie={
      Object {
        "_id": "1",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
  <SingleDisplay
    key="2"
    movie={
      Object {
        "_id": "2",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
  <SingleDisplay
    key="3"
    movie={
      Object {
        "_id": "3",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
  <SingleDisplay
    key="4"
    movie={
      Object {
        "_id": "4",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
  <SingleDisplay
    key="5"
    movie={
      Object {
        "_id": "5",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
  <SingleDisplay
    key="6"
    movie={
      Object {
        "_id": "6",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
  <SingleDisplay
    key="7"
    movie={
      Object {
        "_id": "7",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
  <SingleDisplay
    key="8"
    movie={
      Object {
        "_id": "8",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
  <SingleDisplay
    key="9"
    movie={
      Object {
        "_id": "9",
        "actors": Array [
          "Robert Downey Jr.",
          "Chris Evans",
        ],
        "genre": Array [
          "action",
          "comedy",
        ],
        "poster": "url",
        "thumbsDown": 37,
        "thumbsUp": 196,
        "title": "Avengers: Endgame",
        "year": 2019,
      }
    }
  />
</div>
`;
